pallets/lbp/src/lib.rs

```diff
@@ -0,0 +1,1017 @@
+
+			pool_data.paused = false;
+			<PoolData<T>>::insert(&pool_id, &pool_data);
+
+			Self::deposit_event(Event::Unpaused(who, pool_id));
+			Ok(().into())
+		}
+
+		#[pallet::weight(<T as Config>::WeightInfo::add_liquidity())]
+		#[transactional]
+		pub fn add_liquidity(
+			origin: OriginFor<T>,
+			pool_id: PoolId<T>,
+			amount_a: BalanceOf<T>,
+			amount_b: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			ensure!(<PoolOwner<T>>::contains_key(&pool_id), Error::<T>::PoolNotFound);
+
+			Self::ensure_pool_ownership(&who, &pool_id)?;
+
+			ensure!(
+				!amount_a.is_zero() || !amount_b.is_zero(),
+				Error::<T>::CannotAddZeroLiquidity
+			);
+
+			let (mut reserve_a, mut reserve_b) = Self::pool_balances(&pool_id);
+			let (asset_a, asset_b) = Self::pool_assets(&pool_id);
+
+			let pool_data = Self::pool_data(&pool_id);
+
+			let now = <frame_system::Pallet<T>>::block_number();
+			ensure!(
+				pool_data.start == Zero::zero() || now < pool_data.start,
+				Error::<T>::SaleStarted
+			);
+
+			if !amount_a.is_zero() {
+				ensure!(
+					T::MultiCurrency::free_balance(asset_a, &who) >= amount_a,
+					Error::<T>::InsufficientAssetBalance
+				);
+
+				reserve_a = reserve_a.checked_add(amount_a).ok_or(Error::<T>::LiquidityOverflow)?;
+			}
+
+			if !amount_b.is_zero() {
+				ensure!(
+					T::MultiCurrency::free_balance(asset_b, &who) >= amount_b,
+					Error::<T>::InsufficientAssetBalance
+				);
+
+				reserve_b = reserve_b.checked_add(amount_b).ok_or(Error::<T>::LiquidityOverflow)?;
+			}
+
+			T::MultiCurrency::transfer(asset_a, &who, &pool_id, amount_a)?;
+			T::MultiCurrency::transfer(asset_b, &who, &pool_id, amount_b)?;
+
+			<PoolBalances<T>>::insert(&pool_id, (reserve_a, reserve_b));
+			Self::deposit_event(Event::LiquidityAdded(pool_id, asset_a, asset_b, amount_a, amount_b));
+
+			Ok(().into())
+		}
+
+		#[pallet::weight(<T as Config>::WeightInfo::add_liquidity())]
+		#[transactional]
+		pub fn remove_liquidity(
+			origin: OriginFor<T>,
+			pool_id: PoolId<T>,
+			amount_a: BalanceOf<T>,
+			amount_b: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			ensure!(<PoolOwner<T>>::contains_key(&pool_id), Error::<T>::PoolNotFound);
+
+			Self::ensure_pool_ownership(&who, &pool_id)?;
+
+			ensure!(
+				!amount_a.is_zero() || !amount_b.is_zero(),
+				Error::<T>::CannotRemoveZeroLiquidity
+			);
+
+			let (mut reserve_a, mut reserve_b) = Self::pool_balances(&pool_id);
+			let (asset_a, asset_b) = Self::pool_assets(&pool_id);
+
+			let pool_data = Self::pool_data(&pool_id);
+
+			ensure!(!Self::is_pool_running(&pool_data), Error::<T>::SaleNotEnded);
+
+			if !amount_a.is_zero() {
+				reserve_a = reserve_a.checked_sub(amount_a).ok_or(Error::<T>::LiquidityUnderflow)?;
+			}
+
+			if !amount_b.is_zero() {
+				reserve_b = reserve_b.checked_sub(amount_b).ok_or(Error::<T>::LiquidityUnderflow)?;
+			}
```

Comment:
Same applies here what was mentioned for the add_liquduity. It can reorganized so check are done prior to making any storage calls.